cmd/kubeadm/app/constants/constants.go

```diff
@@ -452,6 +452,39 @@ func EtcdSupportedVersion(versionString string) (*version.Version, error) {
 	return nil, errors.Errorf("unsupported or unknown Kubernetes version(%v)", kubernetesVersion)
 }
 
+// EtcdNearestVersion returns the nearest version to officially supported version of etcd for a specific Kubernetes release
+// If the distance is same, it returns the newer version
+func EtcdNearestVersion(versionString string) (*version.Version, error) {
+	kubernetesVersion, err := version.ParseSemantic(versionString)
+	if err != nil {
+		return nil, err
+	}
+	kubernetesMinorVersion := uint8(kubernetesVersion.Minor())
+
+	key, distance, etcdNearestStringVersion := uint8(0), ^uint8(0), ""
```

Comment:
instead of calculating the distance between "versionString" and the desired version, can we just assume that the user wants a version older or newer than what we have in the map, because EtcdSupportedVersion() already failed.

```
EtcdNearestVersion(versionString string) (*version.Version, error) {
	kubernetesVersion, err := version.ParseSemantic(versionString)
	if err != nil {
		return nil, err
	}

	min, max := ^uint8(0), uint8(0)
	for k := range SupportedEtcdVersion {
		if k < min {
			min = k
		} else if k > max {
			max = k
		}
	}
	
	var nearest string	
	desiredVersion := uint8(kubernetesVersion.Minor())
	if desiredVersion < min {
		nearest = SupportedEtcdVersion[min]
	}
	nearest = SupportedEtcdVersion[max]

	etcdNearestVersion, err := version.ParseSemantic(nearest)
	if err != nil {
		return nil, err
	}
	return etcdNearestVersion, nil
}
```

Reply:
Your suggestion is very good. I have refactored this part of the code :)